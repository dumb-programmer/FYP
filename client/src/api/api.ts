export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
