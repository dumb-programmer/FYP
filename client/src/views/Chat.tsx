import MessageList from "@/components/MessageList";
import { useForm } from "react-hook-form";
import { useQuery } from "react-query";
import { useParams } from "react-router-dom";

export default function Chat() {
    const { chatId } = useParams();
    const { data: messages } = useQuery(`chat-${chatId}`, {
        queryFn: async () => {
            const response = await fetch(`http://localhost:3000/chats/${chatId}/messages`, { mode: "cors", credentials: "include" });
            if (response.ok) {
                return await response.json();
            }
        }
    });
    const { handleSubmit, register } = useForm({
        defaultValues: {
            prompt: ""
        }
    })

    return <div className="flex flex-col h-screen overflow-hidden">
        <h2>{chatId}</h2>
        <div className="flex-1 p-10 flex flex-col gap-10 overflow-y-auto">
            <MessageList messages={messages}/>
        </div>
        <div className="flex justify-center p-5 bg-slate-200 w-full">
            <form className="flex gap-2 w-4/5" onSubmit={handleSubmit(async (data) => {
                const response = await fetch(`http://localhost:3000/chats/${chatId}`, {
                    method: "POST",
                    mode: "cors",
                    credentials: "include",
                    body: JSON.stringify(data),
                    headers: {
                        "Content-Type": "application/json"
                    }
                });
            })}>
                <input className="input input-borderd flex-1" type="text" placeholder="Prompt..." {...register("prompt")} />
                <button className="btn btn-primary">Send</button>
            </form>
        </div>
    </div>;
}