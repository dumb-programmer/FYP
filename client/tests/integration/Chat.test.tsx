import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach } from "vitest";
import Chat from "../../src/views/Chat";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "react-query";
import { getChatName, getMessages, searchMessages } from "../../src/api/api";
import userEvent from "@testing-library/user-event";

vi.mock("../../src/hooks/useSocketContext", async () => {
    return {
        default: () => ({ on: vi.fn(), off: vi.fn() })
    }
})

vi.mock("../../src/api/api", async () => {
    const actual = await vi.importActual("../../src/api/api");

    return {
        ...actual,
        getChatName: vi.fn().mockReturnValue({ ok: true, json: () => ({ name: "Chat 123" }) }),
        searchMessages: vi.fn(),
        getMessages: vi.fn()
    }
})

const queryClient = new QueryClient();
const renderComponent = () => render(
    <MemoryRouter initialEntries={["/chat/123"]}>
        <Routes>
            <Route path="/chat/:chatId" element={
                <QueryClientProvider client={queryClient}>
                    <Chat />
                </QueryClientProvider>} />
        </Routes>
    </MemoryRouter>
);

describe("Chat", () => {
    beforeEach(() => {
        window.IntersectionObserver = vi.fn(() => ({ observe: vi.fn(), disconnect: vi.fn() }))
        HTMLDivElement.prototype.scrollBy = vi.fn()
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("renders correctly", async () => {
        await waitFor(() => {
            const { container } = renderComponent();
            expect(container).toMatchSnapshot();
        })
    });

    it("By default getMessages is called and all messages are shown", async () => {
        await waitFor(() => {
            const { container } = renderComponent();
            expect(container).toMatchSnapshot();
        })

        expect(getMessages).toHaveBeenCalled();
        expect(getMessages).toHaveBeenCalledWith("123", 1);
    });

    it("When the user types a query, searchMessages is called and SearchResults are shown", async () => {
        await waitFor(() => {
            const { container } = renderComponent();
            expect(container).toMatchSnapshot();
        });

        await userEvent.type(screen.getByPlaceholderText(/search/i), "test");

        expect(searchMessages).toHaveBeenCalled();
        expect(searchMessages).toHaveBeenCalledWith("123", "test", 1);
    });
})