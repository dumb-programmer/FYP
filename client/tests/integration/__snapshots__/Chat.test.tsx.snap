// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chat > By default getMessages is called and all messages are shown 1`] = `
<div>
  <div
    class="flex flex-col h-screen overflow-hidden"
  >
    <div
      class="p-5"
    >
      <div
        class="relative"
        data-testid="searchbar"
      >
        <svg
          aria-hidden="true"
          class="absolute mt-[13px] ml-3"
          color="#696969"
          data-slot="icon"
          fill="currentColor"
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fill-rule="evenodd"
          />
        </svg>
        <input
          class="input input-bordered w-full pl-10"
          placeholder="Search..."
          value=""
        />
      </div>
    </div>
    <div
      class="flex-1 p-10 flex flex-col gap-10 overflow-y-auto"
    >
      <div />
    </div>
    <div
      class="flex justify-center p-5 bg-base-100 w-full"
    >
      <form
        class="flex gap-2 w-4/5"
        data-testid="prompt-form"
      >
        <input
          class="input input-bordered flex-1"
          name="prompt"
          placeholder="Prompt..."
          required=""
          type="text"
        />
        <button
          aria-label="send"
          class="btn btn-primary "
        >
          <svg
            aria-hidden="true"
            data-slot="icon"
            fill="currentColor"
            height="20"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.87 2.298a.75.75 0 0 0-.812 1.021L3.39 6.624a1 1 0 0 0 .928.626H8.25a.75.75 0 0 1 0 1.5H4.318a1 1 0 0 0-.927.626l-1.333 3.305a.75.75 0 0 0 .811 1.022 24.89 24.89 0 0 0 11.668-5.115.75.75 0 0 0 0-1.175A24.89 24.89 0 0 0 2.869 2.298Z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Chat > When the user types a query, searchMessages is called and SearchResults are shown 1`] = `
<div>
  <div
    class="flex flex-col h-screen overflow-hidden"
  >
    <div
      class="p-5"
    >
      <div
        class="relative"
        data-testid="searchbar"
      >
        <svg
          aria-hidden="true"
          class="absolute mt-[13px] ml-3"
          color="#696969"
          data-slot="icon"
          fill="currentColor"
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fill-rule="evenodd"
          />
        </svg>
        <input
          class="input input-bordered w-full pl-10"
          placeholder="Search..."
          value=""
        />
      </div>
    </div>
    <div
      class="flex-1 p-10 flex flex-col gap-10 overflow-y-auto"
    >
      <div />
    </div>
    <div
      class="flex justify-center p-5 bg-base-100 w-full"
    >
      <form
        class="flex gap-2 w-4/5"
        data-testid="prompt-form"
      >
        <input
          class="input input-bordered flex-1"
          name="prompt"
          placeholder="Prompt..."
          required=""
          type="text"
        />
        <button
          aria-label="send"
          class="btn btn-primary "
        >
          <svg
            aria-hidden="true"
            data-slot="icon"
            fill="currentColor"
            height="20"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.87 2.298a.75.75 0 0 0-.812 1.021L3.39 6.624a1 1 0 0 0 .928.626H8.25a.75.75 0 0 1 0 1.5H4.318a1 1 0 0 0-.927.626l-1.333 3.305a.75.75 0 0 0 .811 1.022 24.89 24.89 0 0 0 11.668-5.115.75.75 0 0 0 0-1.175A24.89 24.89 0 0 0 2.869 2.298Z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Chat > renders correctly 1`] = `
<div>
  <div
    class="flex flex-col h-screen overflow-hidden"
  >
    <div
      class="p-5"
    >
      <div
        class="relative"
        data-testid="searchbar"
      >
        <svg
          aria-hidden="true"
          class="absolute mt-[13px] ml-3"
          color="#696969"
          data-slot="icon"
          fill="currentColor"
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.965 11.026a5 5 0 1 1 1.06-1.06l2.755 2.754a.75.75 0 1 1-1.06 1.06l-2.755-2.754ZM10.5 7a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Z"
            fill-rule="evenodd"
          />
        </svg>
        <input
          class="input input-bordered w-full pl-10"
          placeholder="Search..."
          value=""
        />
      </div>
    </div>
    <div
      class="flex-1 p-10 flex flex-col gap-10 overflow-y-auto"
    >
      <div />
    </div>
    <div
      class="flex justify-center p-5 bg-base-100 w-full"
    >
      <form
        class="flex gap-2 w-4/5"
        data-testid="prompt-form"
      >
        <input
          class="input input-bordered flex-1"
          name="prompt"
          placeholder="Prompt..."
          required=""
          type="text"
        />
        <button
          aria-label="send"
          class="btn btn-primary "
        >
          <svg
            aria-hidden="true"
            data-slot="icon"
            fill="currentColor"
            height="20"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.87 2.298a.75.75 0 0 0-.812 1.021L3.39 6.624a1 1 0 0 0 .928.626H8.25a.75.75 0 0 1 0 1.5H4.318a1 1 0 0 0-.927.626l-1.333 3.305a.75.75 0 0 0 .811 1.022 24.89 24.89 0 0 0 11.668-5.115.75.75 0 0 0 0-1.175A24.89 24.89 0 0 0 2.869 2.298Z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</div>
`;
